text/0013-prepare-blockbuilder-and-core-runtime-apis-for-mbms.md

```diff
@@ -0,0 +1,127 @@
+# RFC-0013: Prepare `BlockBuilder` and `Core` runtime API for MBMs
+
+|                 |                                                                             |
+| --------------- | --------------------------------------------------------------------------- |
+| **Start Date**  | July 24, 2023 |
+| **Description** | Prepare the `BlockBuilder` and `Core` Runtime APIs for Multi-Block-Migrations.|
+| **Authors**     | Oliver Tale-Yazdi |
+
+## Summary
+
+Introduces breaking changes to the `BlockBuilder` and `Core` runtime APIs and bumps them to version
+5 and 7 respectively. A new function `after_inherents` is added to the `BlockBuilder` runtime API
+and the `initialize_block` function of `Core` is changed to return an enum to indicate what
+extrinsics can be applied.
+
+## Motivation
+
+The original motivation is Multi-Block-Migrations. They require the runtime to be able to tell the
+block builder that it should not attempt to include transactions in the current block. Currently,
+there is no communication possible between runtime and block builder that could achieve this.
+Further, it is necessary to execute runtime logic right after inherent application but still before
+transaction inclusion.
+
+This RFC proposes a way of communication between the runtime and the block builder by changing the
+`initialize_block` function to return a `ExtrinsicInclusionMode` enum. Additionally, an
+`after_inherents` function is introduced that runs after inherent but before transaction
+application.  
+
+## Stakeholders
+
+- Substrate Maintainers: They will have to implement this upstream with all the testing, audit and
+  maintenance burden.
+- Polkadot Runtime developers: They will have to adapt to this breaking change.
+- Polkadot Parachain Teams: They also have to adapt to the breaking changes but then eventually have
+  multi-block migrations available.
+
+## Explanation
+
+The only relevant change is on the node side in the block authoring logic. Any further preparation
+for MBMs can happen entirely on the runtime side with the provided primitives.
+
+All block authors MUST respect the `ExtrinsicInclusionMode` that is returned by `initialize_block`.
+They MUST always invoke `after_inherents` directly after inherent application. The runtime MAY
```

Comment:
what parameters does `after_inherents` take? Is there any way to limit the amount of weight this can spend?

The use-case I'm thinking of is consensus algorithms where the amount of time which can be spent in authoring is variable, e.g. asynchronous backing scenarios where some blocks might be authored in a full 2 seconds and others only in 1 second.

Currently, block builders have the ability to gauge how heavy transactions are, and to stop authoring after submitting enough transactions.

Block builders should be able to limit the amount of time spent in `after_inherents`, although probably not arbitrarily.

Perhaps the `ExtrinsicInclusionMode` should return a minimum amount of weight that must be allotted to the rest of the block, so block builders can decide whether to simply give up authoring or continue.

Reply:
To some extent I'm wondering if it makes sense to flip block-building inside out - in theory we could have a single `fn build_block(InherentData, some_total_weight_or_time_limit)` and supply new transactions through special host functions instead, so the runtime actually decides whether to pull transactions from the pool, etc.

Reply:
Weight is a runtime concept. How much weight you want to spent is also something that the runtime should decide and not the node. All information that you want to pass between initialize and after inherents, can be stored temporarily in the state and doesn't need to be passed to the function.

Regarding driving the block production from inside the runtime. I think there exist an issue or we at least discussed it at some point. The main problem being a panic in a transaction. We can not unwind the panic. We would may need to spawn a new was instance to ensure block production isn't affected by a panic.

Reply:
> How much weight you want to spent is also something that the runtime should decide and not the node.

I disagree - there's no good way for the runtime to know how much time is available for authoring in the general case.